libpod/image/pull.go

```diff
@@ -198,12 +200,74 @@ func (ir *Runtime) pullGoalFromImageReference(ctx context.Context, srcRef types.
 
 // pullImageFromHeuristicSource pulls an image based on inputName, which is heuristically parsed and may involve configured registries.
 // Use pullImageFromReference if the source is known precisely.
-func (ir *Runtime) pullImageFromHeuristicSource(ctx context.Context, inputName string, writer io.Writer, authfile, signaturePolicyPath string, signingOptions SigningOptions, dockerOptions *DockerRegistryOptions, label *string) ([]string, error) {
+func (ir *Runtime) pullImageFromHeuristicSource(ctx context.Context, config *NewImageConfig) ([]string, error) {
 	span, _ := opentracing.StartSpanFromContext(ctx, "pullImageFromHeuristicSource")
 	defer span.Finish()
+	sc := GetSystemContext(config.SignaturePolicyPath, config.Authfile, false)
+	goal, err := ir.getPullGoalFromHeuristicSource(ctx, sc, config.Name)
+	if err != nil {
+		return nil, err
+	}
+	return ir.doPullImage(ctx, sc, *goal, config.Writer, config.SigningOptions, config.DockerOptions, config.Label)
+}
+
+// pullImageFromHeuristicSourceIfNecessary checks an image is out of date with its source,
+// and if so, pulls an updated image.
+// Rather than search for the name in config, use the names of localImg, so we don't have to
+// refind the transport and reconstruct the correct name.
+func (ir *Runtime) pullImageFromHeuristicSourceIfNecessary(ctx context.Context, localImg *storage.Image, config *NewImageConfig) ([]string, error) {
+	span, _ := opentracing.StartSpanFromContext(ctx, "pullImageFromHeuristicSourceIfNecessary")
+	defer span.Finish()
+
+	var inspectErrors *multierror.Error
 
+	sc := GetSystemContext(config.SignaturePolicyPath, config.Authfile, false)
+	var goal *pullGoal
+	for _, name := range localImg.Names {
+		goal, _ = ir.getPullGoalFromHeuristicSource(ctx, sc, name)
+		if goal != nil {
+			break
+		}
+	}
+	if goal == nil {
+		return nil, errors.Errorf("Unable to get pull goal for updating image")
+	}
+
+	found := false
+	for _, imageInfo := range goal.refPairs {
+		if config.Writer != nil && (imageInfo.srcRef.Transport().Name() == DockerTransport || imageInfo.srcRef.Transport().Name() == AtomicTransport) {
```

Comment:
(It would be nice to be able to avoid repetition of this logic… but let’t see the conceptual issues first.)